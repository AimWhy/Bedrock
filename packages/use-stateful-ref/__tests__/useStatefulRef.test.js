import React from "react";
import ReactDOM from "react-dom/";
import { act } from "react-dom/test-utils";

import useStatefulRef from "../src";

jest.spyOn(React, "useState");

let state = undefined;

const setState = jest.fn((newState) => {
  state = newState;
});

React.useState.mockReturnValue([state, setState]);

let statefulRef;
const HookWrapper = ({ value }) => {
  statefulRef = useStatefulRef();
  if (value) {
    statefulRef.current = value;
  }
  return null;
};

describe("useStatefulRef", () => {
  let container;

  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });

  afterEach(() => {
    statefulRef = undefined;
    document.body.removeChild(container);
    container = null;
  });

  test("useStatefulRef is not null", () => {
    expect(useStatefulRef).toBeTruthy();
  });

  it("should call useState", () => {
    act(() => {
      ReactDOM.render(<HookWrapper />, container);
    });

    expect(React.useState).toBeCalled();
  });

  it("should call setState when setting value", () => {
    act(() => {
      ReactDOM.render(<HookWrapper value="Test" />, container);
    });

    expect(setState).toBeCalled();
    expect(statefulRef).toMatchSnapshot();
  });
});
