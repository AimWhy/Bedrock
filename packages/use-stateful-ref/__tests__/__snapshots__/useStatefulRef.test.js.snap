// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStatefulRef should call setState when setting value 1`] = `
Object {
  "current": "Test",
}
`;
